### PR TITLE
Special exit status when no implicated PR was found, retry fetching PRs

### PR DESCRIPTION
**Pull request by gudmdharalds**
Instead of exiting with success (status `0`) when no pull request was found to scan, exit with failure (status `230`). This change is made because exiting indicating success can be confusing to users, but exiting with failure indicates that something might be amiss. 

Also with this change, if no pull request was found to scan, `vip-go-ci` will retry the operation once. Sometimes the GitHub API does not invalidate its cache instantly when a pull request is created, leading to no pull requests being found, but with this change this will hopefully happen less often.

Combined, these two changes will make it clearer to users that no pull request was found when this is truly the case, and also reduce the risk of this unintentionally happening when a pull request has recently been created. 

TODO:
- [X] Change exit status when no implicated PR was found to scan
- [X] Retry fetching PRs 
- [x] Update README
- [X] Changelog entry [#190]
- [x] Check automated unit-tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -651,12 +651,40 @@ php ~/vip-go-ci-tools/vip-go-ci/github-commit-status.php --repo-owner=`repo-owne
...
 	export BUILD_DESCRIPTION="No problems were identified"
-else
+
+elif [ "$VIPGOCI_EXIT_CODE" == "230" ] ; then
+	export BUILD_STATE="failure"
+	export BUILD_DESCRIPTION="Pull request not found for commit"
+
+elif [ "$VIPGOCI_EXIT_CODE" == "249" ] ; then
+	export BUILD_STATE="failure"
+	export BUILD_DESCRIPTION="Build timed out, PR may be too large"
```
Comment (ariskataoka) [inline README.md:666]: ❤️

File: defines.php [hunk 2]
```diff
@@ -75,12 +75,15 @@
...
 
-define( 'VIPGOCI_EXIT_NORMAL',		0 );
-define( 'VIPGOCI_EXIT_EXEC_TIME', 	249 );
-define( 'VIPGOCI_EXIT_CODE_ISSUES',	250 );
-define( 'VIPGOCI_EXIT_SYSTEM_PROBLEM',	251 );
-define( 'VIPGOCI_EXIT_GITHUB_PROBLEM',	252 );
-define( 'VIPGOCI_EXIT_USAGE_ERROR',	253 );
+define( 'VIPGOCI_EXIT_NORMAL',			0 );
+
+define( 'VIPGOCI_EXIT_COMMIT_NOT_PART_OF_PR',	230 );
```
Comment (ariskataoka) [inline defines.php:80]: 👍

**Review by ariskataoka — COMMENTED** [review]
Thanks for implementing that! It seems great.

I would recommend adding tests to cover the scenarios where:

- Cache would be skipped;
- Retries are expected if it's possible;
- Sysexit functions with its new parameters options;

**Comment by gudmdharalds** [comment]
Thanks @ariskataoka for looking into this PR. 

***

Regarding unit-testing the following:

> Cache would be skipped;

The only way I see to do this is to get the printed output from the test while running it. It is certainly possible, but it will take a while to implement. I'm going to move this into another issue so we can have a look into it later, see #164.

> Retries are expected if it's possible;

This is possible, but it would have to be against a special HTTP API, returning different results in first attempt from the second one. We do not have this in the unit-test package currently and implementing it is non-trivial.

> Sysexit functions with its new parameters options;

To the best of my knowledge unit-testing with `exit()` in place is not possible, we would always have to work around it. We would also have to get the printed output to test it. I added this to #164 .

In order to be able to release tomorrow, we will need to postpone work on these items until later. Please indicate if you see any problems with this.

**Comment by gudmdharalds** [comment]
Regarding unit-test for `vipgoci_sysexit()`, see #200 for a possible solution.

**Comment by ariskataoka** [comment]
>The only way I see to do this is to get the printed output from the test while running it. It is certainly possible, but it will take a while to implement. I'm going to move this into another issue so we can have a look into it later, see #164.

I believe we can start thinking of adding tests for the new code implemented from now. And let the #164 issue address unit tests for the legacy that has no unit tests implemented by now? What do you think?

>This is possible, but it would have to be against a special HTTP API, returning different results in first attempt from the second one. We do not have this in the unit-test package currently and implementing it is non-trivial.

This makes me wonder once more about how we've been implementing our tests. For example, in this case, in a scenario where we were to mock APIs, the unit tests would be less complex to implement. 🤔

**Comment by gudmdharalds** [comment]
> I believe we can start thinking of adding tests for the new code implemented from now. And let the #164 issue address unit tests for the legacy that has no unit tests implemented by now? What do you think?

Yes, I agree that this is a good goal. 

However, I think we should keep in mind exceptions. For example, when tests are very complex to implement and take a long time, while testing very limited aspect of the functionality. I'm happy to discuss this in more detail.

> This makes me wonder once more about how we've been implementing our tests. For example, in this case, in a scenario where we were to mock APIs, the unit tests would be less complex to implement.

This is what I thought too. Being able to implement mock APIs easily would be of benefit.

**Pull request merged**